mmagic/models/editors/animatediff/attention_3d.py

```diff
@@ -0,0 +1,699 @@
+                self.norm2(hidden_states, timestep)
+                if self.use_ada_layer_norm else self.norm2(hidden_states))
+            hidden_states = (
+                self.attn2(
+                    norm_hidden_states,
+                    encoder_hidden_states=encoder_hidden_states,
+                    attention_mask=attention_mask) + hidden_states)
+
+        # Feed-forward
+        hidden_states = self.ff(self.norm3(hidden_states)) + hidden_states
+
+        # Temporal-Attention
+        if self.unet_use_temporal_attention:
+            d = hidden_states.shape[1]
+            hidden_states = rearrange(
+                hidden_states, '(b f) d c -> (b d) f c', f=video_length)
+            norm_hidden_states = (
+                self.norm_temp(hidden_states, timestep)
+                if self.use_ada_layer_norm else self.norm_temp(hidden_states))
+            hidden_states = self.attn_temp(norm_hidden_states) + hidden_states
+            hidden_states = rearrange(
+                hidden_states, '(b d) f c -> (b f) d c', d=d)
+
+        return hidden_states
+
+
+class CrossAttention(nn.Module):
+    r"""
+    A cross attention layer.
+
+    Parameters:
+        query_dim (`int`): The number of channels in the query.
+        cross_attention_dim (`int`, *optional*):
+            The number of channels in the encoder_hidden_states.
+            If not given, defaults to `query_dim`.
+        heads (`int`,  *optional*, defaults to 8):
+        The number of heads to use for multi-head attention.
+        dim_head (`int`,  *optional*, defaults to 64):
+        The number of channels in each head.
+        dropout (`float`, *optional*, defaults to 0.0):
+        The dropout probability to use.
+        bias (`bool`, *optional*, defaults to False):
+            Set to `True` for the query, key, and value linear
+            layers to contain a bias parameter.
+    """
+
+    def __init__(
+        self,
+        query_dim: int,
+        cross_attention_dim: Optional[int] = None,
+        heads: int = 8,
+        dim_head: int = 64,
+        dropout: float = 0.0,
+        bias=False,
+        upcast_attention: bool = False,
+        upcast_softmax: bool = False,
+        added_kv_proj_dim: Optional[int] = None,
+        norm_num_groups: Optional[int] = None,
+    ):
+        super().__init__()
+        inner_dim = dim_head * heads
+        cross_attention_dim = cross_attention_dim \
+            if cross_attention_dim is not None else query_dim
+        self.upcast_attention = upcast_attention
+        self.upcast_softmax = upcast_softmax
+
+        self.scale = dim_head**-0.5
+
+        self.heads = heads
+        # for slice_size > 0 the attention score computation
+        # is split across the batch axis to save memory
+        # You can set slice_size with `set_attention_slice`
+        self.sliceable_head_dim = heads
+        self._slice_size = None
+        self._use_memory_efficient_attention_xformers = True
```

Comment:
this lead to the error below.
![image](https://github.com/open-mmlab/mmagic/assets/12782558/aea20480-5a1f-4191-9bd8-600fe88ddf54)

Reply:
I've modified this to run inference without xformers. But it's still highly recommended to install xformers for better memory cost.